src/liballoc/collections/vec_deque.rs

```diff
@@ -2170,12 +2170,20 @@ impl<'a, T> Iterator for Iter<'a, T> {
         back.iter().fold(accum, &mut f)
     }
 
-    fn try_fold<B, F, R>(&mut self, init: B, mut f: F) -> R where
-        Self: Sized, F: FnMut(B, Self::Item) -> R, R: Try<Ok=B>
+    fn try_fold<B, F, R>(&mut self, init: B, mut f: F) -> R
+    where
+        Self: Sized,
+        F: FnMut(B, Self::Item) -> R,
+        R: Try<Ok = B>,
     {
         let (front, back) = RingSlices::ring_slices(self.ring, self.head, self.tail);
-        let accum = front.iter().try_fold(init, &mut f)?;
-        back.iter().try_fold(accum, &mut f)
+        let mut front_iter = front.iter();
+        let res = front_iter.try_fold(init, &mut f);
+        self.tail = self.ring.len() - front_iter.len();
+        let mut back_iter = back.iter();
+        let res = back_iter.try_rfold(res?, f);
+        self.tail = back.len() - back_iter.len();
```

Comment:
I'm concerned here again.  Suppose we did 
```rust
let mut v = VecDeque::with_capacity(7);
v.push_back(1);
v.push_back(1);
v.push_back(1);
v.pop_front();
v.pop_front();
```

Then the backing store is 

```rust
_ _ 1 _ _ _ _ _
head = 2
tail = 3
```

so `(front, back) == ([1], [])`, and iterating will result in setting `tail = 0 - 0`, leaving an iterator that thinks it contains 6 elements, and safe code would then be exposed to uninitialized values.  (Not to mention that the iter implements `FusedIterator`, so it's not supposed to return non-`None` after returning `None` once.